Run TestAOTAutograd test suite with cache

jamesjwu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #128583
* #128335
* #128329
* __->__ #128222

This diff introduces AOTAutogradTestWithCache, which runs AOTAutogradTests with both dynamo and AOTAutogradCache. 

To do this, for any verify_aot_autograd() calls in the original tests, we run compiled_f an extra time. We also turn on a new strict mode that throws any time a cache is missed due to weird reasons, like BypassAOTAutogradCache or FxGraphCacheMiss. 

We use a mocked version of FXGraphCache to decrease the number of variables for these tests. The normal tests in test_aot_autograd_cache.py will still run with FXGraphCache. I might change my mind and unmock these in the future. 


In total, 87 of the tests pass naturally. None of the tests fail in non strict cache mode, so the cache never crashes, it just misses more often than we'd like. The remaining 27 tests fail due to relatively simple (though not necessarily easy to fix) reasons. I'll fix the remaining test failures in the next few PRs.

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Rebase onto main

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Fix bug in xfail

[ghstack-poisoned]

jamesjwu commented on test/functorch/test_aotdispatch.py:
```diff
@@ -285,24 +287,20 @@ def is_in_base(t, maybe_tensors):
         def wrapper(self, *args, **kwargs):
-            fn = func
             if isinstance(self, TestAOTAutogradWithDynamo):
-                if xfail:
```
The previous xfail actually wasn't doing what I want: in fact it was just skipping the test and passing it. Because in python 3 xfail doesn't work on functions but on test cases, you have to set it on the class before self is known (which is dumb because skipTest still works?). 

So I'm using xfail_inherited_tests to handle the xfail tests in the dynamo suite instead.

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Code reviews

[ghstack-poisoned]

jamesjwu committed:
Fixes

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Rebase for clean tests

[ghstack-poisoned]

bdhirsh commented on test/functorch/test_aotdispatch.py:
```diff
@@ -429,13 +428,46 @@ def verify_aot_autograd(
                         # mutate it.
-                        ref_o.mul_(2)
-                        test_o.mul_(2)
+                        ref_o.add(2)
```
we probably want to keep the `mul_(2)` from the line above, instead of doing an out-of-place add. The test is trying to confirm that we can mutate the output of a compiled region without making autograd unhappy / raising an error

jamesjwu replied:
Hmm ok, I chose add because it was reversible but I can probably just recreate a new ref_out each time

bdhirsh replied:
oh `add_` is fine too - what I meant is that the test is trying to inplace-mutate the data of the output of the compiled function, to ensure that autograd won't error on the mutation (so we lose test coverage if we chaneg it to be an out-of-place op like `test_o.add()`)

jamesjwu replied:
Oh I did not realize .`add()` was out of place instead of `add_`. Will change!

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Code review

[ghstack-poisoned]

bdhirsh reviewed (approved):
lgtm!

jamesjwu commented:
@pytorchbot rebase

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Update on "Run TestAOTAutograd test suite with cache"


This diff introduces AOTAutogradTestWithCache, which runs AOTAutogradTests with both dynamo and AOTAutogradCache. 

To do this, for any verify_aot_autograd() calls in the original tests, we run compiled_f an extra time. We also turn on a new strict mode that throws any time a cache is missed due to weird reasons, like BypassAOTAutogradCache or FxGraphCacheMiss. 

We use a mocked version of FXGraphCache to decrease the number of variables for these tests. The normal tests in test_aot_autograd_cache.py will still run with FXGraphCache. I might change my mind and unmock these in the future. 


In total, 87 of the tests pass naturally. None of the tests fail in non strict cache mode, so the cache never crashes, it just misses more often than we'd like. The remaining 27 tests fail due to relatively simple (though not necessarily easy to fix) reasons. I'll fix the remaining test failures in the next few PRs.



[ghstack-poisoned]

jamesjwu commented:
@pytorchbot rebase

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
One more rebase

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu commented:
@pytorchbot merge

jamesjwu commented:
@pytorchbot merge
